src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -491,5 +495,102 @@ private static CastResult TryGet(nuint source, nuint target)
 
             return ref Unbox_Helper(toTypeHnd, obj);
         }
+
+        internal struct ArrayElement
+        {
+            public object? Value;
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        private static ref object? ThrowArrayMismatchException()
+        {
+            throw new ArrayTypeMismatchException();
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        private static ref object? LdelemaRef(Array array, int index, void* type)
+        {
+            ArrayElement[] arr = Unsafe.As<ArrayElement[]>(array);
+
+            // this will throw appropriate exceptions if array is null or access is out of range.
+            nuint elementType = RuntimeHelpers.GetMethodTable(arr)->ElementType;
+            ref object? element = ref arr[index].Value;
+
+            if (elementType != (nuint)type)
+                goto throwArrayMismatch;
+
+            return ref element;
+
+            throwArrayMismatch:
+                return ref ThrowArrayMismatchException();
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        private static void StelemRef(Array array, int index, object? obj)
+        {
+            ArrayElement[] arr = Unsafe.As<ArrayElement[]>(array);
+
+            // this will throw appropriate exceptions if array is null or access is out of range.
+            nuint elementType = RuntimeHelpers.GetMethodTable(arr)->ElementType;
+            ref object? element = ref arr[index].Value;
+
+            if (obj == null)
+                goto assigningNull;
+
+            if (elementType != (nuint)RuntimeHelpers.GetMethodTable(obj))
+                goto notExactMatch;
+
+            doWrite:
+                WriteBarrier(ref element, obj);
+                return;
+
+            assigningNull:
+                element = null;
+                return;
+
+            notExactMatch:
+                if (elementType == (nuint)pObjMt)
+                    goto doWrite;
+
+            StelemRef_Helper(ref element, elementType, obj);
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        private static void StelemRef_Helper(ref object? element, nuint elementType, object obj)
+        {
+            CastResult result = TryGet((nuint)RuntimeHelpers.GetMethodTable(obj), (nuint)elementType);
+            if (result == CastResult.CanCast)
+            {
+                WriteBarrier(ref element, obj);
+                return;
+            }
+
+            StelemRef_Helper_Slow(ref element, elementType, obj);
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        private static void StelemRef_Helper_Slow(ref object? element, nuint elementType, object obj)
```

Comment:
```suggestion
        private static void StelemRef_Helper_NoCacheLookup(ref object? element, nuint elementType, object obj)
```

So that it has consistent naming with the casting helpers?